bitcoin/tx.c

```diff
@@ -80,29 +80,21 @@ static void push_witnesses(const struct bitcoin_tx *tx,
 
 static void push_tx(const struct bitcoin_tx *tx,
 		   void (*push)(const void *, size_t, void *), void *pushp,
-		   bool extended)
+		   bool include_witness)
 {
 	varint_t i;
-	u8 flag = 0;
 
 	push_le32(tx->version, push, pushp);
+        if (include_witness && !uses_witness(tx))
+            include_witness = false;
 
-	if (extended) {
-		u8 marker;
+	if (include_witness) {
+		u8 marker = 0;
+	        u8 flag = SEGREGATED_WITNESS_FLAG;
 		/* BIP 144 */
 		/* marker 	char 	Must be zero */
 		/* flag 	char 	Must be nonzero */
-		marker = 0;
 		push(&marker, 1, pushp);
-		/* BIP 141: The flag MUST be a 1-byte non-zero
```

Comment:
I prefer to switch the logic around, to make it more future-proof.

Keep flag = 0;  Then if we're in bip144 mode, or in SEGREGATED_WITNESS_FLAG iff uses(witness(tx)).  That leaves room for future checks if other flags exist.  Then push the flag iff it's non-zero, keeping the BIP141 quote.